vm/src/serializer.h

```diff
@@ -0,0 +1,68 @@
+#ifndef BORGES_SERIALIZER_H
+#define BORGES_SERIALIZER_H
+
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+namespace Borges {
+
+class Serializer {
+public:
+    Serializer(std::vector<double>& b);
+    void serialize(double x);
+    template <typename T>
+    void serialize(const std::shared_ptr<T>& p);
+
+private:
+    std::vector<double>& buffer_;
+    std::unordered_map<std::shared_ptr<void>, std::size_t> ptr_to_index_map_;
+};
+
+template <typename T>
+void
+Serializer::serialize(const std::shared_ptr<T>& p)
+{
+    auto it = ptr_to_index_map_.find(p);
+    if (it == ptr_to_index_map_.end()) {
+      serialize(*this, it->second);
+      return;
+    }
+    std::size_t i = ptr_to_index_map_.size();
+    serialize(i);
+    serialize(*p);
+    ptr_to_index_map_.insert(std::make_pair(std::shared_ptr<void>(p), i));
+}
```

Comment:
why do you serialize the size of the index map ?

Reply:
The fact that I used i and not n as the name for the variable should have been a hint that this is an index, not a size. I use the size of the map for the value of the index because I need a unique index that has not been used before. If that isn't clear from this code, I will add a comment to this effect.

Reply:
i read it as an index, but noticed the value, which made me think. I usually use -1 to indicate that the value has not been used or 0(zero) if all valid indices are > 0. otherwise, indeed, please comment.